### ISSUE TITLE
optimizers wishlist

### ISSUE THREAD

**Issue by skeydan** [mlverse/torch#147]
`Adam` (unsurprisingly I guess)

thanks!

**Comment by skeydan** [comment]
adding `LBFGS` just out of enthusiasm ;-) 
(not using anywhere yet but suspect it could be real good for some types of data)

**Comment by dirkschumacher** [comment]
Can I add LBFGS or is anyone already working on this?

**Comment by dirkschumacher** [comment]
Why don't you use the C++ implementations of the optimizers?

**Comment by dfalbel** [comment]
I am not working on LBFGS. Pinging @krzjoa as he worked on many other implementations.

The main reason for not using the C++ implementations was that I wanted to allow extensions in the R side, and It would tricky to call custom R functions if using the C++ class. Other reasons are 

- many of them are simple enough to be reimplemented. 
- the C++ implementation doesn't support parameter groups.

That said, we can have optimizers that instantiate the C++ object when initializing and then having the `step` method calling the C++ step method.

**Comment by krzjoa** [comment]
I haven't started working on LBFGS yet, so feel free to implement it, @dirkschumacher :wink:

**Comment by dirkschumacher** [comment]
Ok, given the complexity of LBFGS I would rather use an existing implementation than rewriting it in R. I will take a look and come back with questions :)

**Comment by krzjoa** [comment]
In this place I would like to propose an initial road map of `torch` optimizers.
In my opinion, our first goal could be to implement all optimizers , that are curently present in [PyTorch](https://pytorch.org/docs/stable/optim.html) and [Keras](https://keras.io/api/optimizers/).

**PyTorch optimizers**:

- [X] Adadelta
- [X] Adagrad
- [X] Adam
- [ ] AdamW
- [ ] SparseAdam
- [ ] Adamax
- [X] ASGD
- [ ] LBFGS
- [X] RMSprop
- [X] Rprop
- [X] SGD

**Keras optimizers** (non occurring in PyTorch):

- [ ] Nadam
- [ ] Ftrl

Additionaly, there is a couple of [MXNet optimizers](https://mxnet.apache.org/versions/1.7/api/python/docs/tutorials/packages/optimizer/index.html) which didn't appear above.

Some other fancier optimizers could be implemented in a separate library like [pytorch-optimizer](https://github.com/jettify/pytorch-optimizer).

**Comment by dirkschumacher** [comment]
Maybe it would make sense to extract the optimizers into a different package anyways and reexport them in the `torch`package.

**Comment by krzjoa** [comment]
I understand your intention, but personally I would not recommend that (i.e. extracting and reexporting whole optim module). Optimizers cannot work without `torch`, so we'd create a cyclic graph of dependencies. 😉

**Comment by dirkschumacher** [comment]
Yeah, I agree. Cyclic dependencies are considered harmful :)

**Comment by dfalbel** [comment]
Sounds good to me! 
+1 for a new package for fancier optimizers! that would be really cool

**Comment by egillax** [comment]
Hi @dfalbel ,

> The main reason for not using the C++ implementations was that I wanted to allow extensions in the R side, and It would tricky to call custom R functions if using the C++ class. Other reasons are
> 
>     many of them are simple enough to be reimplemented.
>     the C++ implementation doesn't support parameter groups.
> 
> That said, we can have optimizers that instantiate the C++ object when initializing and then having the step method calling the C++ step method.

Would it be possible to have optimizers that use the C++ objects? For example Adam or AdamW which are both available in libtorch. I believe this could speed up my training quite a bit since it seems that the optimizer$step() is a bottleneck compared to pytorch. Also if I'm reading the cpp docs [correctly](https://pytorch.org/cppdocs/api/classtorch_1_1optim_1_1_optimizer_param_group.html#_CPPv4N5torch5optim19OptimizerParamGroupE) it seems C++ now supports parameter groups.

**Comment by dfalbel** [comment]
Yes, In theory that's possible. I'll draft something in that direction and post here. 

Ideally I'd really want to figure what's slowing down optimizers in R compared to PyTorch, because being able to inherit from other optimizers and etc is really useful for research and verification. Still I don't see what could be causing this, maybe the way we keep state? Anyway, I think having those C++ based optimizers is good.

**Comment by dfalbel** [comment]
@egillax Here's a POC binding directly to the C++ LibTorch optimizers:

https://github.com/dfalbel/torchoptx

I didn't benchmark at all, but curious to see if that's much faster than the R based optimizers.
Currently only SGD and Adam are supported, but shouldn't be a lot of work to add the others.
We would also need to figure out how to support serialization.

**Comment by egillax** [comment]
Hi @dfalbel,

That was quick! But I can't open the repo, is it by any chance private?

**Comment by dfalbel** [comment]
@egillax Nice, this sounds promising! I think we should be able to achieve very similar from R - specially on GPU where the operations are non-blocking. 

This will probably make more difference the more parameters the network has though.
